### PR TITLE
Assigenment2 - Trevor Felchlin 

### PR DESCRIPTION
**Pull request by Tfelchlin13**

### CODE REVIEW

**Comment by connerhanson** [comment]
I'll review this one.

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 1]
```diff
@@ -26,14 +26,25 @@ public void Error_WithData_LogsMessage()
...
+
+    }
+    [TestMethod]
+    public void CheckingForClassNameProperty_WhenInstantiatingTestLogger_ShouldReturnTrue()
+    {
+        // Arrange
+        BaseLogger logger = new TestLogger(); // Use TestLogger instead of BaseLogger
+
+        // Act & Assert
+        Assert.IsTrue(typeof(BaseLogger).GetProperty("ClassName") != null);
```
Comment (connerhanson) [inline Logger.Tests/BaseLoggerMixinsTests.cs:44]: I like how you combined the act and assert functions into one line.

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 2]
```diff
@@ -26,14 +26,25 @@ public void Error_WithData_LogsMessage()
...
     }
 
+
```
Comment (connerhanson) [inline Logger.Tests/BaseLoggerMixinsTests.cs:47]: Since tests should be done for each situation you should test for each method in your BaseLoggerMixin.cs.  I would recommend writing two tests for each, with one checking null results and the other testing if the method does the correct work for each call.  You can pretty much just copy the two error tests given, substituting the error with the other methods (warning, information, debug).

**Review by connerhanson — COMMENTED** [review]
Overall everything looks great, adding a few more tests for BaseLoggerMixin.cs may be good, though the method from that class all look great.  Everything else looks good!


    There is an existing BaseLogger class. It needs an auto property to hold a class name. This property should be set in the LogFactory using an object initializer. ✔
    Create a FileLogger that derives from BaseLogger. It should take in a path to a file to write the log message to. When its Log method is called, it should append messages on their own line in the file. The output should include all of the following:
        The current date/time ✔
        The name of the class that created the logger ✔
        The log level ✔
        The message ✔
        The format may vary, but an example might look like this "10/7/2019 12:38:59 AM FileLoggerTests Warning: Test message"
    The LogFactory should be updated with a new method ConfigureFileLogger. This should take in a file path and store it in a private member. It should use this when instantiating a new FileLogger in its CreateLogger method. ❌, after reviewing the code review on my pull-request I realized I missed this, ConfigureFileLogger must be called inside the the CreateLogger method. 

    If the file logger has not be configured in the LogFactory, its CreateLogger method should return null. ✔
    Inside of BaseLoggerMixins implement extension methods on BaseLogger for
        Error, ✔
        Warning, ✔
        Information, and ✔
        Debug. ✔ 
Each of these methods should take in a string for the message, as well as a parameter array of arguments for the message. Each of these extension methods is expected to be a shortcut for calling the BaseLogger.Log method, by automatically supplying the appropriate LogLevel. These methods should throw an exception if the BaseLogger parameter is null. There are a couple example unit tests to get you started.
    Use the nameof() operator when identifying the class name to the logger ✔
    Ensure you turn on Warnings as Errors (TreatWarningsAsErrors) ✔
    Ensure that you turn on code analysis (EnableNETAnalyzers) ✔
    Ensure that you turn on CodeAnalysisTreatWarningsAsErrors ✔
    Ensure that you turn on EnforceCodeStyleInBuild ❌
    Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
    Turn on Nullability (Nullable) ✔
    All of the above should be unit tested. - Need some tests for BaseLoggerMixin.cs methods

File: Logger/LogFactory.cs [hunk 3]
```diff
@@ -1,10 +1,28 @@
...
+﻿using System;
+#nullable enable
+using System.IO;
+namespace Logger;
 
 public class LogFactory
 {
-    public BaseLogger CreateLogger(string className)
+    private string? _FilePath;
+    public BaseLogger? CreateLogger(string className)
```
Comment (connerhanson) [inline Logger/LogFactory.cs:9]: Use the configureFileLogger method inside this method.

File: Logger.Tests/BaseLoggerMixinsTests.cs (hunk 2 above)
Comment (Tfelchlin13) [reply Logger.Tests/BaseLoggerMixinsTests.cs:150]: Tests added i struggled with test for a bit but i sat down and really learned them better

File: Logger/LogFactory.cs (hunk 3 above)
Comment (Tfelchlin13) [reply Logger/LogFactory.cs:9]: not sure exactly what you mean

File: Logger.Tests/BaseLoggerMixinsTests.cs [hunk 4]
```diff
@@ -23,15 +56,94 @@ public void Error_WithNullLogger_ThrowsException()
...
-        Assert.AreEqual("Message 42", logger.LoggedMessages[0].Message);
+        Assert.AreEqual("BaseLogger Error: Message 42", logger.LoggedMessages[0].Message);
+
+    }
+    [TestMethod]
+    public void CheckingForClassNameProperty_WhenInstantiatingTestLogger_ShouldReturnTrue()
+    { 
+
+        // Act & Assert
+        Assert.IsTrue(typeof(BaseLogger).GetProperty("ClassName") != null);
```
Comment (BenjaminMichaelis) [inline Logger.Tests/BaseLoggerMixinsTests.cs:75]: 1. I would assert not null here instead of is true
2. I'm confused on what this is testing and why this is necessary?

File: Logger.Tests/FileLoggerTests.cs [hunk 5]
```diff
@@ -1,9 +1,107 @@
...
+            // Act
+            fileLogger.Log(LogLevel.Information, logMessage);
+
+            // Assert
+            var fileContent = File.ReadAllText(_testFilePath);
+            Assert.IsTrue(fileContent.Contains("TestClass"));
+        }
+
+        [TestMethod]
+        public void Log_WithNullClassName_SuccessfullyAppendsUnknownClassName()
```
Comment (BenjaminMichaelis) [inline Logger.Tests/FileLoggerTests.cs:65]: Why would we expect that it can have an unknown class name?

File: Logger.Tests/Logger.Tests.csproj [hunk 6]
```diff
@@ -1,16 +1,22 @@
...
+		<TargetFramework>net8.0</TargetFramework>
 		<IsPackable>false</IsPackable>
-		<LangVersion>11.0</LangVersion>
+		<LangVersion>12.0</LangVersion>
+    <Nullable>enable</Nullable>
+    <CodeAnalysisTreatWarningsAsErrors>true</CodeAnalysisTreatWarningsAsErrors>
+    <EnableNETAnalyzers>true</EnableNETAnalyzers>
+    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
+    <EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
 	</PropertyGroup>
```
Comment (BenjaminMichaelis) [inline Logger.Tests/Logger.Tests.csproj:12]:
```suggestion
		<LangVersion>12.0</LangVersion>
		<Nullable>enable</Nullable>
		<CodeAnalysisTreatWarningsAsErrors>true</CodeAnalysisTreatWarningsAsErrors>
		<EnableNETAnalyzers>true</EnableNETAnalyzers>
		<TreatWarningsAsErrors>true</TreatWarningsAsErrors>
		<EnforceCodeStyleInBuild>true</EnforceCodeStyleInBuild>
	</PropertyGroup>
```
nit: spacing should be standard

File: Logger.Tests/Logger.Tests.csproj [hunk 7]
```diff
@@ -1,16 +1,22 @@
...
-		<PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.2.0" />
-		<PackageReference Include="MSTest.TestAdapter" Version="1.4.0" />
-		<PackageReference Include="MSTest.TestFramework" Version="1.4.0" />
-		<PackageReference Include="coverlet.collector" Version="1.0.1" />
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.8.0" />
+    <PackageReference Include="MSTest.TestAdapter" Version="3.1.1" />
+    <PackageReference Include="MSTest.TestFramework" Version="3.1.1" />
+    <PackageReference Include="coverlet.collector" Version="6.0.0">
+    </PackageReference>
 	</ItemGroup>
```
Comment (BenjaminMichaelis) [inline Logger.Tests/Logger.Tests.csproj:20]:
```suggestion
		<PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.8.0" />
		<PackageReference Include="MSTest.TestAdapter" Version="3.1.1" />
		<PackageReference Include="MSTest.TestFramework" Version="3.1.1" />
		<PackageReference Include="coverlet.collector" Version="6.0.0"/>
	</ItemGroup>
```
nit: spacing should be consistent

File: Logger/BaseLogger.cs [hunk 8]
```diff
@@ -1,7 +1,40 @@
-﻿namespace Logger;
+﻿using System;
+using System.Globalization;
+using System.IO;
+#nullable enable
```
Comment (BenjaminMichaelis) [inline Logger/BaseLogger.cs:4]: Why is this necessary

File: Logger/BaseLogger.cs [hunk 9]
```diff
@@ -1,7 +1,40 @@
...
 
 public abstract class BaseLogger
 {
     public abstract void Log(LogLevel logLevel, string message);
+
+    public string? ClassName { get; set; }
+
+    
+}
+public class FileLogger : BaseLogger
```
Comment (BenjaminMichaelis) [inline Logger/BaseLogger.cs:15]: This should be in it's own file

File: Logger/BaseLogger.cs [hunk 10]
```diff
@@ -1,7 +1,40 @@
...
+using System.Globalization;
+using System.IO;
+#nullable enable
+namespace Logger;
 
 public abstract class BaseLogger
 {
     public abstract void Log(LogLevel logLevel, string message);
+
+    public string? ClassName { get; set; }
```
Comment (BenjaminMichaelis) [inline Logger/BaseLogger.cs:11]: Why is this nullable?

File: Logger/BaseLogger.cs [hunk 11]
```diff
@@ -1,7 +1,40 @@
...
+
+    
+}
+public class FileLogger : BaseLogger
+{
+    private readonly string filePath;
+
+    public FileLogger(string filePath)
+    {
+        this.filePath = filePath;
```
Comment (BenjaminMichaelis) [inline Logger/BaseLogger.cs:21]: `this` is not necessary here

File: Logger/BaseLogger.cs [hunk 12]
```diff
@@ -1,7 +1,40 @@
...
 {
     public abstract void Log(LogLevel logLevel, string message);
+
+    public string? ClassName { get; set; }
+
+    
+}
+public class FileLogger : BaseLogger
+{
+    private readonly string filePath;
```
Comment (BenjaminMichaelis) [inline Logger/BaseLogger.cs:17]: We should use a property over a field here

File: Logger/BaseLoggerMixins.cs [hunk 13]
```diff
@@ -1,6 +1,41 @@
...
+using System.Globalization;
 
-public static class BaseLoggerMixins
-{
+namespace Logger;
 
-}
+    public static class BaseLoggerMixins
+    {
+        public static void Error(this BaseLogger? logger, string message, params object[] args)
```
Comment (BenjaminMichaelis) [inline Logger/BaseLoggerMixins.cs:8]: Why is logger nullable? I expect to call this message, a logger is needed.

File: Logger/BaseLoggerMixins.cs [hunk 14]
```diff
@@ -1,6 +1,41 @@
...
+        {
+            LogWithLevel(logger, LogLevel.Information, message, args);
+        }
+
+        public static void Debug(this BaseLogger? logger, string message, params object[] args)
+        {
+            LogWithLevel(logger, LogLevel.Debug, message, args);
+        }
+
+        public static void LogWithLevel(BaseLogger? logger, LogLevel logLevel, string message, params object[] args)
```
Comment (BenjaminMichaelis) [inline Logger/BaseLoggerMixins.cs:28]: The naming here is weird, this seems to focus more on formatting the string rather than log

File: Logger/LogFactory.cs (hunk 3 above)
Comment (BenjaminMichaelis) [inline Logger/LogFactory.cs:9]: These should be properties

File: Logger/LogFactory.cs [hunk 15]
```diff
@@ -1,10 +1,28 @@
...
+        if (_FilePath == null)
+        {
 
-        return null;
+            return null;
+        }
+        else {
+
+          BaseLogger logger = new FileLogger(_FilePath);
+            return logger;
```
Comment (BenjaminMichaelis) [inline Logger/LogFactory.cs:19]: This can be done in a single line
```suggestion
          return new FileLogger(_FilePath);
```

File: Logger/LogLevel.cs [hunk 16]
```diff
@@ -1,5 +1,5 @@
 ﻿namespace Logger;
-
+#nullable enable
```
Comment (BenjaminMichaelis) [inline Logger/LogLevel.cs:2]: Why is this necessary here?

File: Logger/BaseLogger.cs (hunk 8 above)
Comment (Tfelchlin13) [reply Logger/BaseLogger.cs:4]: Ah i forgot to take that one out i was doing it wrong at first. thanks for the catch.

File: Logger/LogLevel.cs (hunk 16 above)
Comment (Tfelchlin13) [inline Logger/LogLevel.cs:2]:
```suggestion
```

File: Logger/BaseLogger.cs (hunk 8 above)
Comment (Tfelchlin13) [inline Logger/BaseLogger.cs:4]:
```suggestion
```
